File: backend/app/schemas/quiz.py
"""Pydantic schemas for quizzes and attempts."""
from __future__ import annotations

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, field_validator


class QuizCreate(BaseModel):
    unit_id: int
    num_questions: int = 10
    difficulty: str = "medium"


class QuizResponse(BaseModel):
    id: int
    unit_id: int
    questions: list[dict]
    difficulty: str
    created_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class QuizAttemptRequest(BaseModel):
    # `None` entries represent unanswered questions (skipped/out of range).
    answers: list[Optional[int]]


class QuizAttemptResult(BaseModel):
    score: int
    total: int
    percentage: float
    results: list[dict]
    # G13 (Phase 85): XP granted for a passing attempt (0 when none).
    xp_awarded: int = 0


class HistoryItem(BaseModel):
    id: int
    quiz_id: int
    answers: list[int]
    score: Optional[int]
    total_questions: Optional[int]
    created_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/services/api/types/quiz.ts
// SyllabusAI quiz types.
// Extracted from services/api.ts (F8 split).

// ----- SyllabusAI quizzes -----
export interface QuizQuestion {
  question: string;
  options: string[];
  correct_index: number;
  explanation: string;
}

export interface Quiz {
  id: number;
  unit_id: number;
  questions: QuizQuestion[];
  difficulty: string;
  created_at?: string | null;
}

export interface QuizAttemptResultItem {
  question_index: number;
  selected: number | null;
  correct: boolean;
  correct_index: number;
  explanation: string;
}

export interface QuizAttemptResult {
  score: number;
  total: number;
  percentage: number;
  results: QuizAttemptResultItem[];
  // G13 (Phase 85): XP granted for a passing attempt (0 when none).
  xp_awarded?: number;
}

export interface QuizHistoryItem {
  id: number;
  quiz_id: number;
  answers: number[];
  score: number | null;
  total_questions: number | null;
  created_at?: string | null;
}

export interface QuizAnalytics {
  avg_score: number;
  best_score: number;
  total_attempts: number;
  per_unit: Record<number, number>;
}
